src/charm.py

```diff
@@ -8,18 +8,136 @@
+            f"http://kratos.{self.model.name}.svc.cluster.local:4433/self-service/login/browser"
+        )
+        self._kratos_session_url = (
+            f"http://kratos.{self.model.name}.svc.cluster.local:4433/sessions/whoami"
+        )
+
+        self.service_patcher = KubernetesServicePatch(
+            self, [("oathkeeper-api", OATHKEEPER_API_PORT)]
+        )
+
+        self.framework.observe(self.on.oathkeeper_pebble_ready, self._on_oathkeeper_pebble_ready)
+
+    @property
+    def _oathkeeper_layer(self) -> Layer:
+        """Returns a pre-configured Pebble layer."""
+        layer_config = {
+            "summary": "oathkeeper-operator layer",
+            "description": "pebble config layer for oathkeeper-operator",
+            "services": {
+                self._container_name: {
+                    "override": "replace",
+                    "summary": "Oathkeeper Operator layer",
+                    "command": f"oathkeeper serve -c {self._oathkeeper_config_path}",
+                    "startup": "enabled",
+                }
+            },
+            "checks": {
+                "alive": {
+                    "override": "replace",
+                    "http": {"url": f"http://localhost:{OATHKEEPER_API_PORT}/health/alive"},
+                },
+                "ready": {
+                    "override": "replace",
+                    "http": {"url": f"http://localhost:{OATHKEEPER_API_PORT}/health/ready"},
+                },
+            },
+        }
+        return Layer(layer_config)
+
+    @property
+    def _oathkeeper_service_is_created(self) -> bool:
+        try:
+            self._container.get_service(self._container_name)
+        except (ModelError, RuntimeError):
+            return False
+        return True
+
+    def _render_access_rules_file(self) -> str:
+        """Render the access rules file."""
+        with open("templates/access-rules.yaml.j2", "r") as file:
+            template = Template(file.read())
+
+        rendered = template.render(
+            kratos_login_url=self._kratos_login_url,
+            return_to="http://default-url.com",
+        )
+        return rendered
+
+    def _render_conf_file(self) -> str:
+        """Render the Oathkeeper configuration file."""
+        with open("templates/oathkeeper.yaml.j2", "r") as file:
+            template = Template(file.read())
+
+        rendered = template.render(
+            kratos_session_url=self._kratos_session_url,
+            kratos_login_url=self._kratos_login_url,
+        )
+        return rendered
+
+    def _on_oathkeeper_pebble_ready(self, event: PebbleReadyEvent) -> None:
+        """Event Handler for pebble ready event."""
+        if not self._container.can_connect():
+            event.defer()
+            logger.info("Cannot connect to Oathkeeper container. Deferring the event.")
+            self.unit.status = WaitingStatus("Waiting to connect to Oathkeeper container")
+            return
```

Comment:
Why do we need to check this? If the pebble ready event is fired is it not safe to assume that we can connect to the container?

Reply:
According to [these](https://juju.is/docs/sdk/container-name-pebble-ready-event) docs:

> it’s essential to, also in pebble-ready event handlers, to guard with Container.can_connect() to check that the socket is up

At the time the pebble event is fired it should be able to connect to the container, but since the event can be deferred, I think it makes sense to check it.